File: problems-adapt/0701-0800/0745_prefix-suffix-word-search/solution_dual_trie.go
package main

// Two tries, one word list per node: a prefix trie spelling every word
// forward and a suffix trie spelling every word reversed, so a suffix
// reads down it front to back. Words are inserted in index order, so
// every node's list ascends, and bestMatch walks pref down the first
// trie and suff reversed down the second, then merges the two hit nodes'
// lists from their tails -- the first equal pair is the largest shared
// index, and a walk that falls off its trie means no word matches that
// half, answering -1.
type trieNode struct {
	children [26]*trieNode
	indices  []int
}

type PrefixSuffixIndex struct {
	prefixes *trieNode
	suffixes *trieNode
}

func NewPrefixSuffixIndexTyped(words []string) *PrefixSuffixIndex {
	design := &PrefixSuffixIndex{prefixes: newTrieNode(), suffixes: newTrieNode()}
	for index, word := range words {
		node := design.prefixes
		for position := 0; position < len(word); position++ {
			node = node.child(word[position])
			node.indices = append(node.indices, index)
		}
		node = design.suffixes
		for position := len(word) - 1; position >= 0; position-- {
			node = node.child(word[position])
			node.indices = append(node.indices, index)
		}
	}
	return design
}

func newTrieNode() *trieNode {
	return &trieNode{}
}

// child returns the slot for character, creating the node on first use.
func (node *trieNode) child(character byte) *trieNode {
	slot := int(character) - 97
	if node.children[slot] == nil {
		node.children[slot] = &trieNode{}
	}
	return node.children[slot]
}

func (design *PrefixSuffixIndex) bestMatch(pref string, suff string) int {
	forward := design.walkForward(pref)
	if forward == nil {
		return -1
	}
	backward := design.walkBackward(suff)
	if backward == nil {
		return -1
	}
	front := forward.indices
	back := backward.indices
	for i, j := len(front)-1, len(back)-1; i >= 0 && j >= 0; {
		if front[i] == back[j] {
			return front[i]
		}
		if front[i] > back[j] {
			i--
		} else {
			j--
		}
	}
	return -1
}

// walkForward descends the prefix trie one node per character of pref;
// nil as soon as a slot is empty.
func (design *PrefixSuffixIndex) walkForward(pref string) *trieNode {
	node := design.prefixes
	for index := 0; index < len(pref); index++ {
		node = node.children[int(pref[index])-97]
		if node == nil {
			return nil
		}
	}
	return node
}

// walkBackward descends the suffix trie, whose edges spell the reversed
// words, so suff is consumed from its end; nil as soon as a slot is empty.
func (design *PrefixSuffixIndex) walkBackward(suff string) *trieNode {
	node := design.suffixes
	for index := len(suff) - 1; index >= 0; index-- {
		node = node.children[int(suff[index])-97]
		if node == nil {
			return nil
		}
	}
	return node
}

File: problems-originals/0701-0800/0745_prefix-and-suffix-search/solution_dual_trie.cpp
#include <string>
#include <vector>

// Two tries, one word list per node: a prefix trie spelling every word
// forward and a suffix trie spelling every word reversed, so a suffix
// reads down it front to back. Words are inserted in index order, so
// every node's list ascends, and f() walks pref down the first
// trie and suff reversed down the second, then merges the two hit nodes'
// lists from their tails -- the first equal pair is the largest shared
// index, and a walk that falls off its trie means no word matches that
// half, answering -1.
class WordFilter {
  public:
    WordFilter(vector<string> words) {
        for (int index = 0; index < (int)words.size(); ++index) {
            const string &word = words[index];
            Node *node = &prefixes;
            for (char character : word) {
                node = node->step(character);
                node->indices.push_back(index);
            }
            node = &suffixes;
            for (int position = (int)word.size() - 1; position >= 0; --position) {
                node = node->step(word[position]);
                node->indices.push_back(index);
            }
        }
    }

    int f(string pref, string suff) {
        const Node *forward = walk(pref);
        if (forward == nullptr) {
            return -1;
        }
        const Node *backward = walkReversed(suff);
        if (backward == nullptr) {
            return -1;
        }
        const vector<int> &front = forward->indices;
        const vector<int> &back = backward->indices;
        int i = (int)front.size() - 1;
        int j = (int)back.size() - 1;
        while (i >= 0 && j >= 0) {
            if (front[i] == back[j]) {
                return front[i];
            }
            if (front[i] > back[j]) {
                --i;
            } else {
                --j;
            }
        }
        return -1;
    }

  private:
    struct Node {
        Node *children[26] = {};
        vector<int> indices;

        // Returns the child slot for character, creating it on first use.
        Node *step(char character) {
            Node *&slot = children[character - 'a'];
            if (slot == nullptr) {
                slot = new Node;
            }
            return slot;
        }
    };

    // Walks pref down the prefix trie; nullptr as soon as a slot is empty.
    const Node *walk(const string &pref) const {
        const Node *node = &prefixes;
        for (char character : pref) {
            node = node->children[character - 'a'];
            if (node == nullptr) {
                return nullptr;
            }
        }
        return node;
    }

    // Walks suff down the suffix trie, whose edges spell the reversed
    // words, so the characters are consumed from the end; nullptr as soon
    // as a slot is empty.
    const Node *walkReversed(const string &suff) const {
        const Node *node = &suffixes;
        for (int index = (int)suff.size() - 1; index >= 0; --index) {
            node = node->children[suff[index] - 'a'];
            if (node == nullptr) {
                return nullptr;
            }
        }
        return node;
    }

    Node prefixes;
    Node suffixes;
};
